src/service/service.rs

```diff
@@ -0,0 +1,43 @@
+use std::future::Future;
+
+/// An asynchronous function from a `Request` to a `Response`.
+///
+/// The `Service` trait is a simplified interface making it easy to write
+/// network applications in a modular and reusable way, decoupled from the
+/// underlying protocol.
+///
+/// # Functional
+///
+/// A `Service` is a function of a `Request`. It immediately returns a
+/// `Future` representing the eventual completion of processing the
+/// request. The actual request processing may happen at any time in the
+/// future, on any thread or executor. The processing may depend on calling
+/// other services. At some point in the future, the processing will complete,
+/// and the `Future` will resolve to a response or error.
+///
+/// At a high level, the `Service::call` function represents an RPC request. The
+/// `Service` value can be a server or a client.
+pub trait Service<Request> {
+    /// Responses given by the service.
+    type Response;
+
+    /// Errors produced by the service.
+    type Error;
+
+    /// The future response value.
+    type Future: Future<Output = Result<Self::Response, Self::Error>>;
+
+    /// Process the request and return the response asynchronously.
+    ///
+    /// This function is expected to be callable off task. As such,
+    /// implementations should take care to not call `poll_ready`.
+    ///
+    /// Before dispatching a request, `poll_ready` must be called and return
+    /// `Poll::Ready(Ok(()))`.
+    ///
+    /// # Panics
+    ///
+    /// Implementations are permitted to panic if `call` is invoked without
+    /// obtaining `Poll::Ready(Ok(()))` from `poll_ready`.
+    fn call(&mut self, req: Request) -> Self::Future;
```

Comment:
Do we want to preserve the `&mut self` here? I know that's how tower's worked, but I think within hyper at least pretty much everything internally would work better with a shared reference so it can be used from multiple tasks without cloning.

Reply:
I think it's a valid question, I'd be happy to read more if you're interested in writing up a proposal to change it. I'm tempted to suggest we have the conversation in a separate issue, to not bog down the the work from Tomek.

Reply:
It makes sense for immutability to be the default, and if mutable state is required, one needs to utilize an interior mutability pattern.

I'd be happy to implement it and even write the proposal if @sfackler doesn't feel like it, but we should definitely leave it out of scope of this MR.